pkg/consolegraphql/cache/cache.go

```diff
@@ -0,0 +1,294 @@
+/*
+* Copyright 2019, EnMasse authors.
+* License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+package cache
+
+import (
+	"fmt"
+	"github.com/hashicorp/go-memdb"
+	"strings"
+)
+
+// ObjectFilters filter the objects return from a cache.
+// return:
+// - match, true if the object should be included in the results
+// - cont, true if the traverse should continue
+// - e in the event of a error preventing evaluation of the filter.
+type ObjectFilter = func(interface{}) (match bool, cont bool, e error)
+
+// ObjectMutator is used with Cache.Update function to mutate existing
+// object(s) in the cache within a transaction. The ObjectMutator function
+// receives the current value it uses to return a replacement  Returning
+// nil causes no change to be made for this value. Remaining objects will
+// continue to be considered for update.
+//
+// It is illegal for the updater function to mutate the object's key(s).
+type ObjectMutator = func(current interface{}) (replacement interface{}, e error)
+
+func And(filters... ObjectFilter) ObjectFilter {
+	return func(o interface{}) (bool, bool, error) {
+		c := true
+		for _, filter := range filters {
+			if filter != nil {
+				fr, fc, e := filter(o)
+				if e != nil {
+					return false, false, e
+				}
+
+				if !fc {
+					c = false
+				}
+				if !fr {
+					return false, fc, nil
+				}
+			}
+		}
+
+		return true, c, nil
+	}
+}
+
+type Cache interface {
+	Init(indexSpecifier ...IndexSpecifier) error
+
+	Add(objs ...interface{}) error
+	Update(mutator ObjectMutator, objs ...interface{}) error
+	Delete(objs ...interface{}) error
+	DeleteByPrefix(idxName, keyPrefix string) error
+
+	Get(idxName string, keyPrefix string, filter ObjectFilter) ([]interface{}, error)
+	Dump() error
+
+	GetKeyCreator(idxName string) (KeyCreator, error)
+}
+
+type MemdbCache struct {
```

Comment:
I think ideally we should be able to set a size limitation on this cache so that we don't run into OOM situations in systems with many addresses.

Also, if bounded, there should be a replacement policy.

Reply:
@lulf unfortunately memdb doesn't have the ability to constrain it own size, nor the ability to demand load.   With the current implementation console-server has to be sized sufficiently large to accommodate addresss+connections+links.  

my idea (long term) is to swap out memdb with another key value store implementation with the ability to page to disk and give console-server a non-persistent transient volume.  That way it would spill out to disk if/when it needed to and this would be transparent to the application.  I know BDB has suitable properties. There may be more modern implementations with better Go bindings.

for now, console-server's memory needs to be configurable for its CR.